### ISSUE TITLE
Typescript: jsonParseReviver - current json object isn't supported

### ISSUE THREAD

**Issue by MirKml** [RSuter/NSwag#1729]
In current Typescript client, it's possible to use own reviver for JSON.parse(json, reviver) call.

    let resultData200 = _responseText === "" ? null : 
      JSON.parse(_responseText, this.jsonParseReviver);

But current implementation doesn't allow use context of reviver, which is current parsed json - see [JSON.parse documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse#Using_the_reviver_parameter). Important is: _Then it is called, with the object containing the property being processed as 'this'_.

I propose this one

    let currentClient = this;
    let resultData200 = _responseText === "" ? null : 
      JSON.parse(_responseText, (key, value) => currentClient.jsonParseReviver(key, value, this));

**Comment by MirKml** [comment]
No, see the spec for JSON.Parse - this - in context of JSON.Parse function itself - is property of current parsed JSON, which is being processed. JSON.Parse change context of 'this' in same way as e.g. jquery. Calling `currentClient.jsonParseReviver(key, value, this)` means currentClient.jsonParseReviver(key, value, processedJsonProperty).

**Comment by RSuter** [comment]
Ok i understand, is this correctly working with a ts lambda? Shoudnt you use a function for the correct this?

**Comment by RSuter** [comment]
Cant you set it this way? https://github.com/RSuter/NSwag/wiki/SwaggerToTypeScriptClientGenerator#add-a-custom-jsonparse-reviver-function

**Comment by MirKml** [comment]
> Ok i understand, is this correctly working with a ts lambda? Shoudnt you use a function for the correct this?

I think it works, I tested only mentioned snippet on TS playgroud, it's compiled to normal ES5 closure with preserved 'this' context.

> Cant you set it this way? https://github.com/RSuter/NSwag/wiki/SwaggerToTypeScriptClientGenerator#add-a-custom-jsonparse-reviver-function

I know about this one, for most of simple cases like multiple values, rename values etc. it's sufficient, but this isn't sufficient for my case, I need to access _processedJsonProperty_, for resolving some special type of information from current processed json. It's possible only through 'this' json property from JSON.Parse reviver context.

**Comment by RSuter** [comment]
![image](https://user-images.githubusercontent.com/2603405/48429802-1df7d480-e76e-11e8-9ab0-89f4f854a5e2.png)

**Comment by RSuter** [comment]
Is this `this` parameter even supported by all browsers?
